File: binomial.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

struct term {
    int num;
    string val;
};

vector<vector<int>> binomial (int size) {
  vector<vector<int>> tri (size, vector<int> (size));

  for (int i = 0; i < size; ++i) {
      tri[i][0] = tri[i][i] = 1;

      for (int j = 1; j < i; ++j) {
          tri[i][j] = tri[i-1][j-1] + tri[i -1][j];
      }
  }
  return tri;
}

void expand (string expr) {
    const int up = expr.find("^");
    string num = expr.substr(0, up), exx = expr.substr(up + 1);
    int ex = stoi(exx);

    int size = ex + 1;
    vector<term> code (size);
    vector<vector<int>> tri = binomial (size);

    for (int i = 0; i < size; i++) {
        int mul = tri[size-1][i];
        // int nu1 = pow(t1.num, ex-i), nu2 = pow(t2.num, i);
        // int total = mul * nu1 * nu2;
        //
        // cout << total;
        //
        // if (t1.val != "") {
        //     if (ex-i > 0) {
        //       cout << t1.val << "^";
        //       cout << ex-i;
        //     }
        // }

        // cout << "+";
        // cout << mul << " " << nu1 << ' ' << nu2  << " :: " << total << "\n";
    }
}

int main () {

    //expand("(x+1)^2");      // returns "x^2+2x+1"
    //expand("(p-1)^3");      // returns "p^3-3p^2+3p-1"
    //expand("(2f+4)^6");     // returns "64f^6+768f^5+3840f^4+10240f^3+15360f^2+12288f+4096"
    //expand("(-2a-4)^0");    // returns "1"
    //expand("(-12t+43)^2");  // returns "144t^2-1032t+1849"
    //expand("(r+0)^203");    // returns "r^203"
    //expand("(-x-1)^2");     // returns "x^2+2x+1

    // (a + b)^5 = a^5 + 5 * a^4 * b^1 + 10 * a^3 b^2 + 10a^2 b^3 + 5a^1 b^4 + a^0 b^5;

    string expr = "(2f+4)^6";

    const int up = expr.find("^");
    string num = expr.substr(0, up), exx = expr.substr(up + 1);
    int ex = stoi(exx);


    // cout << num;

}

File: python/binomial.py

def getterm(str) :
    num = ''.join([ch for ch in str if ch.isdigit()])
    value = 1 if num == '' else int(num)
    sign = -1 if '-' in str else 1
    var = ''.join([ch for ch in str if ch.isalpha()])

    return (value * sign, var)

def binomial(exp) :
    tri = [[1 for _ in range(exp + 1)] for _ in range(exp + 1)]

    for i in range(exp + 1) :
        for j in range(1, i) :
            tri[i][j] = tri[i-1][j] + tri[i-1][j-1]
    return tri

def formexponent (var, ex) :
    match ex :
        case 0 : return ''
        case 1 : return var

    return var + '^' + str(ex)

def expand(expr):
    mid = expr.find("^")
    exp = int(expr[mid+1:])
    term = expr[1:mid-1]
    bino = binomial(exp)
    res = ''
    # if exp == '0' : return '1'
    # if exp == '1' : return term
    for i in range(len(term)) :
        ch = term[i]
        if ch == '+' or ch == '-' :
            v0 = getterm(term[:i]), getterm(term[i:])
            break
    # print('[', v0, ']')

    for i in range(len(bino[exp])) :
        sig = [exp - i, i]
        mult, term = bino[exp][i], ''
#     let mul = (0..2).fold(binom[exp][i].clone(), |mul, x| mul * BigInt::from(v0[x].0).pow(sig[x] as u32));
#     let term = (0..2).filter(|&x| v0[x].1 != None).map(|x| formexponent(v0[x].1.unwrap(), sig[x])).collect::<String>();
        for j in range(2) :
            mult *= pow(v0[j][0], sig[j])
            if v0[j][1] != '' :
                term += formexponent(v0[j][1], sig[j])

        if mult != 0 :
            if i > 0 and mult > 0 :
                res += '+'
            if mult > 1 or mult < -1 :
                res += str(mult)
            elif term == '' :
                res += str(mult)

            res += term

        # print('[', mult, ']', '[', term, ']=>', res)
    #     print('[', mult, ']', '[', term, ']')
    # print(res)
    return res

def assert_equals(a,b) :
    if a != b :
        print("error got : ", a, 'expect : ', b )

assert_equals(expand("(x+1)^0"), "1")
assert_equals(expand("(x-1)^0"), "1")
assert_equals(expand("(7x-7)^0"), "1")
assert_equals(expand("(-7x-7)^0"), "1")
assert_equals(expand("(x+1)^1"), "x+1")
assert_equals(expand("(x+1)^2"), "x^2+2x+1")

assert_equals(expand("(x-1)^1"), "x-1")
assert_equals(expand("(x-1)^2"), "x^2-2x+1")
assert_equals(expand("(2x-3)^3"), "8x^3-36x^2+54x-27")
assert_equals(expand("(5m+3)^4"), "625m^4+1500m^3+1350m^2+540m+81")

# assert_equals(expand("(-2k-3)^3"), "-8k^3-36k^2-54k-27")
#
# assert_equals(expand("(-5m+3)^4"), "625m^4-1500m^3+1350m^2-540m+81")
